### PR TITLE
Add `.exact`, `.upToNextMajor`, and `.upToNextMinor` options to `CompatibleXcodeVersions`

### PR DESCRIPTION
**Pull request by ezraberch**
Resolves #3919

### Short description 📝

This PR adds new options to `CompatibleXcodeVersions` to allow non-exact specification of supported Xcode versions. These new options — `.exact`, `.upToNextMajor`,  `.upToNextMinor` — work as they do in SPM.

If current syntax is used, `.exact` will be used. I'm not sure if this is the best approach long-term, but it maintains current functionality so there is no breaking change.

### How to test the changes locally 🧐

Use the new options in your Config.swift. For example,

```swift
let config = Config(
  compatibleXcodeVersions: [.upToNextMajor("13.2"), "12.5.1"],
  generationOptions: []
)
```

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 1]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
...
 
+    case exact(Version)
+
+    case upToNextMajor(Version)
+
+    case upToNextMinor(Version)
+
     /// List of versions that are supported by the project.
-    case list([String])
+    case list([CompatibleXcodeVersions])
```
Comment (luispadron) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:15]: Don't think it makes sense to change this right?

Wouldn't this allow:

```
compatabileXcodes: [.upToNextMajor("1.0.0"), .upToNextMinor("1.0.0")]
```

File: Sources/TuistGenerator/Linter/EnvironmentLinter.swift [hunk 2]
```diff
@@ -36,18 +36,14 @@ public class EnvironmentLinter: EnvironmentLinting {
...
         }
 
         let version = xcode.infoPlist.version
 
-        if !compatibleVersions.contains(version) {
-            let versions = compatibleVersions.joined(separator: ", ")
+        if !config.compatibleXcodeVersions.isCompatible(versionString: version) {
+            let versions = config.compatibleXcodeVersions
             let message =
                 "The project, which only supports the versions of Xcode \(versions), is not compatible with your selected version of Xcode, \(version)"
```
Comment (luispadron) [inline Sources/TuistGenerator/Linter/EnvironmentLinter.swift:48]: If were in here, would be sweet to have the error be something like:

> The selected Xcode version is x.x.x which is not compatible with this projects Xcode version requirement of 1.0.0 up to 2.0.0

> The selected Xcode version is x.x.x which is not compatible with this projects Xcode version requirement of 1.0.0

> The selected Xcode version is x.x.x which is not compatible with this projects Xcode version requirements of 1.0.0, 2.1.0, 3.0.0

**Review by luispadron — APPROVED** [review]
This is a great addition, been wanting something like this for one of our projects at work.

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift (hunk 1 above)
Comment (ezraberch) [reply Sources/ProjectDescription/CompatibleXcodeVersions.swift:15]: Is there a problem with allowing that? It's likely useless, but not invalid.

The purpose of allowing this form is something like the above example:

```swift
compatibleXcodeVersions: [.upToNextMajor("13.2"), "12.5.1"]
```

File: Sources/TuistGenerator/Linter/EnvironmentLinter.swift (hunk 2 above)
Comment (ezraberch) [reply Sources/TuistGenerator/Linter/EnvironmentLinter.swift:48]: Sure, the error message can be improved.

I don't think saying "up to 2.0.0" (for `.upToNextMajor("1.0")`) is a good idea, as that could be misinterpreted to mean 2.0.0 is valid. And, of course, we don't know what the last version before the next major will be. I'll see what I can come up with.

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 3]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
     /// The project supports all Xcode versions.
     case all
 
+    case exact(Version)
+
+    case upToNextMajor(Version)
+
+    case upToNextMinor(Version)
```
Comment (danyf90) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:12]: Could you please add a brief `///` documentation to each? 🙏

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 4]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
...
     /// List of versions that are supported by the project.
-    case list([String])
+    case list([CompatibleXcodeVersions])
 
     // MARK: - ExpressibleByArrayLiteral
 
-    public init(arrayLiteral elements: [String]) {
-        self = .list(elements)
+    public init(arrayLiteral elements: [CompatibleXcodeVersions]) {
+        self = .list(elements.map { $0 })
```
Comment (danyf90) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:20]: Any reason to not be just
```suggestion
        self = .list(elements)
```
?

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 5]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
...
     }
 
-    public init(arrayLiteral elements: String...) {
-        self = .list(elements)
+    public init(arrayLiteral elements: CompatibleXcodeVersions...) {
+        self = .list(elements.map { $0 })
```
Comment (danyf90) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:24]:
```suggestion
        self = .list(elements)
```

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 6]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
...
+    enum CodingKeys: String, CodingKey {
         case type
         case value
     }
 
     // MARK: - ExpressibleByStringInterpolation
 
     public init(stringLiteral value: String) {
-        self = .list([value])
+        self = .exact("\(value)")
```
Comment (danyf90) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:35]: I find this less surprising
```suggestion
        self = .exact(Version(stringLiteral: value))
```

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift [hunk 7]
```diff
@@ -5,55 +5,76 @@ public enum CompatibleXcodeVersions: ExpressibleByArrayLiteral, ExpressibleByStr
...
     }
 
     // MARK: - Codable
 
     public func encode(to encoder: Encoder) throws {
```
Comment (danyf90) [inline Sources/ProjectDescription/CompatibleXcodeVersions.swift:40]: Don't we have automatic synthesize since 5.5? We can probably drop the whole custom Codable conformance

File: Sources/ProjectDescription/Version.swift [hunk 8]
```diff
@@ -105,11 +105,11 @@ extension Version {
             .split(separator: ".", maxSplits: 2, omittingEmptySubsequences: false)
             .map(String.init).compactMap { Int($0) }.filter { $0 >= 0 }
 
-        guard requiredComponents.count == 3 else { return nil }
+        guard requiredComponents.count > 0 else { return nil }
```
Comment (danyf90) [inline Sources/ProjectDescription/Version.swift:108]:
```suggestion
        guard !requiredComponents.isEmpty else { return nil }
```

File: Sources/ProjectDescription/Version.swift [hunk 9]
```diff
@@ -105,11 +105,11 @@ extension Version {
...
 
         major = requiredComponents[0]
-        minor = requiredComponents[1]
-        patch = requiredComponents[2]
+        minor = requiredComponents.count > 1 ? requiredComponents[1] : 0
+        patch = requiredComponents.count > 2 ? requiredComponents[2] : 0
```
Comment (danyf90) [inline Sources/ProjectDescription/Version.swift:112]: I think it better reflects the logic:
```suggestion
        minor = requiredComponents.count >= 2 ? requiredComponents[1] : 0
        patch = requiredComponents.count >= 3 ? requiredComponents[2] : 0
```

File: Sources/TuistGraph/Models/CompatibleXcodeVersions.swift [hunk 10]
```diff
@@ -1,20 +1,70 @@
...
+    public func isCompatible(versionString: String) -> Bool {
+        let xCodeVersion: Version = "\(versionString)"
+
+        switch self {
+        case .all:
+            return true
+        case let .exact(version):
+            return version == xCodeVersion
+        case let .upToNextMajor(version):
+            return version == xCodeVersion || (version.major == xCodeVersion.major && xCodeVersion > version)
```
Comment (danyf90) [inline Sources/TuistGraph/Models/CompatibleXcodeVersions.swift:29]: Shouldn't be `>=`? Also, let's try to always put the variable as first operand of the comparison. Same for the upToNextMinor case
```suggestion
            return xCodeVersion == version || (xCodeVersion.major == version.major  && xCodeVersion >= version)
```

File: Sources/TuistGraph/Models/CompatibleXcodeVersions.swift [hunk 11]
```diff
@@ -1,20 +1,70 @@
...
+        case let .upToNextMinor(version):
+            return version == xCodeVersion ||
+                (version.major == xCodeVersion.major && version.minor == xCodeVersion.minor && xCodeVersion > version)
+        case let .list(versions):
+            return versions.map { $0.isCompatible(versionString: versionString) }.contains(true)
```
Comment (danyf90) [inline Sources/TuistGraph/Models/CompatibleXcodeVersions.swift:34]:
```suggestion
            return versions.contains { $0.isCompatible(versionString: versionString) }
```

File: Sources/TuistGraph/Models/CompatibleXcodeVersions.swift [hunk 12]
```diff
@@ -1,20 +1,70 @@
...
+        }
+    }
+
+    // MARK: - ExpressibleByStringInterpolation
+
+    public init(stringLiteral value: String) {
+        self = .exact("\(value)")
```
Comment (danyf90) [inline Sources/TuistGraph/Models/CompatibleXcodeVersions.swift:41]: I find this less surprising, same for the others below
```suggestion
        self = .exact(Version(stringLiteral: value))
```

File: Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift [hunk 13]
```diff
@@ -20,4 +20,22 @@ final class CompatibleXcodeVersionsTests: XCTestCase {
...
+
+    func test_codable_when_upToNext() {
+        // Given
+        let subject = CompatibleXcodeVersions.upToNextMajor("13.2")
+        let subject2 = CompatibleXcodeVersions.upToNextMinor("13.2")
+
+        // Then
+        XCTAssertCodable(subject)
+        XCTAssertCodable(subject2)
+    }
```
Comment (danyf90) [inline Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift:40]: Can you add also `isCompatible` tests, covering the various edge cases?

**Review by danyf90 — CHANGES_REQUESTED** [review]

File: Sources/TuistGraph/Models/CompatibleXcodeVersions.swift (hunk 10 above)
Comment (ezraberch) [reply Sources/TuistGraph/Models/CompatibleXcodeVersions.swift]: If they're equal it will short-circuit due to the first clause, so that part won't be executed.

File: Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift (hunk 13 above)
Comment (ezraberch) [reply Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift:40]: I'm not sure what you're looking for here; `isCompatible` is being tested in the `EnvironmentLinterTests` which were added.

Also, now that `Codable` conformance is being synthesized, I don't think these tests serve any purpose. The `ProjectDescription` version of `CompatibleXcodeVersions` is just a struct with cases and initializers now. Would you object to just removing this file?

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift (hunk 7 above)
Comment (ezraberch) [reply Sources/ProjectDescription/CompatibleXcodeVersions.swift]: This has the small side effect of slightly changing the output of `tuist dump`. Let me know if you want this listed as a breaking change.

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift (hunk 1 above)
Comment (luispadron) [reply Sources/ProjectDescription/CompatibleXcodeVersions.swift:18]: Its not a problem since it will resolve correctly and I don't imagine it'll be a common usage

File: Sources/TuistGenerator/Linter/EnvironmentLinter.swift (hunk 2 above)
Comment (luispadron) [reply Sources/TuistGenerator/Linter/EnvironmentLinter.swift:48]: That works for me, and yeah I noticed the possibility of misinterpretation there with the plain English. You could also just display it in swift range format? Ex: `x.x.x..<x.x.y` which Swift developers may be more familiar with (and I think SPM does it like this)

File: Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift (hunk 13 above)
Comment (luispadron) [reply Tests/ProjectDescriptionTests/CompatibleXcodeVersionsTests.swift:40]: Yeah I personally don't see the point in testing `Coddle` IMO. But I know the project has these tests, would be in favor of removing these going forward.

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift (hunk 7 above)
Comment (danyf90) [reply Sources/ProjectDescription/CompatibleXcodeVersions.swift]: Maybe we can remove all the not required custom codable in 3.0? @fortmarek @pepicrft

File: Sources/TuistGraph/Models/CompatibleXcodeVersions.swift (hunk 10 above)
Comment (danyf90) [reply Sources/TuistGraph/Models/CompatibleXcodeVersions.swift]: True, I would rather remove the first == to make the logic clearer

**Review by danyf90 — APPROVED** [review]

**Review by pepicrft — APPROVED** [review]

File: Sources/ProjectDescription/CompatibleXcodeVersions.swift (hunk 7 above)
Comment (fortmarek) [reply Sources/ProjectDescription/CompatibleXcodeVersions.swift]: `tuist dump` output is not really meant to be stable, so I think it's fine to make it breaking now.

File: Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    func test_isCompatible_when_upToNextMajor() {
+        // Given
+        let subject = CompatibleXcodeVersions.upToNextMajor("13.2")
+
+        // Then
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2"))
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2.0"))
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2.2"))
+        XCTAssertTrue(subject.isCompatible(versionString: "13.3.0"))
+        XCTAssertFalse(subject.isCompatible(versionString: "14.2.0"))
```
Comment (danyf90) [inline Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift:55]: I would add text for boundary (i.e. should return false for 14.0.0)

File: Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    }
```
Comment (danyf90) [inline Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift:56]: I would add test that it return false for:

12.3.0
13.0.0
14.0.0

File: Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        // Given
+        let subject = CompatibleXcodeVersions.upToNextMinor("13.2")
+
+        // Then
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2"))
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2.0"))
+        XCTAssertTrue(subject.isCompatible(versionString: "13.2.2"))
+        XCTAssertFalse(subject.isCompatible(versionString: "13.3.0"))
+        XCTAssertFalse(subject.isCompatible(versionString: "14.2.0"))
+    }
```
Comment (danyf90) [inline Tests/TuistGraphTests/Models/CompatibleXcodeVersionsTests.swift:68]: I would add test that it return false for:

12.2.0
13.0.0

**Review by danyf90 — CHANGES_REQUESTED** [review]

**Review by danyf90 — APPROVED** [review]
